app/referee/job_submitter.py

```diff
@@ -25,78 +26,116 @@ def resubmit_check_for_submission_runs_job(scheduler_submission_script):
     return jobid_sub_bytes.decode("utf-8")
 
 
-def start_cluster(config_file, preload_file, worker_outfile, min_workers, max_workers):
+def start_cluster(config_file, preload_file, worker_outfile):
     with open(config_file, encoding="utf-8") as f:
         config = yaml.safe_load(f)
+        print(config)
+        print(config_file)
+
+    job_extra = [
+        f"--output={worker_outfile}",
+        "--open-mode=append",
+    ]
+
     cluster = SLURMCluster(
         extra=[
             f"--preload {preload_file}",
         ],
-        job_extra=[
-            f"--output={worker_outfile}",
-            "--open-mode=append",
-        ],
-        **config["jobqueue"]["slurm"],
+        job_extra=job_extra,
+        **config["jobqueue"]["slurm"]["cluster_settings"],
     )
-    cluster.adapt(
-        minimum_jobs=min_workers,
-        maximum_jobs=max_workers,
+    logger.debug(
+        "Min: %d; Max: %d",
+        config["jobqueue"]["slurm"]["adapt_settings"]["minimum_jobs"],
+        config["jobqueue"]["slurm"]["adapt_settings"]["maximum_jobs"],
     )
+    cluster.adapt(**config["jobqueue"]["slurm"]["adapt_settings"])
     return cluster
 
 
+def set_submission_run_status(submission_run, status):
+    submission_run.status = status
+    submission_run.save(update_fields=["status"])
+
+
+def job_submitter_alive(start_time, check_interval, job_lifetime):
+    return time.time() - start_time + (3 * check_interval) < job_lifetime
+
+
 def check_for_submission_runs(start_time, client, check_interval, job_lifetime):
     n = 0
     logger.debug(
         "Checking for submissions every %d seconds over %d seconds",
         check_interval,
         job_lifetime,
     )
-    while time.time() - start_time + (1.5 * check_interval) < job_lifetime:
-        logger.debug("Checking for submission runs n=%d", n)
-        for run in SubmissionRun.objects.filter(status=Status.PENDING_REMOTE):
-            logger.debug("Added run=%d", run.id)
-            run.status = Status.PENDING
-            run.save(update_fields=["status"])
-            rt.submit_submission_run(client, run)
 
-        time.sleep(check_interval)
+    while job_submitter_alive(start_time, check_interval, job_lifetime):
+        logger.debug("Checking for submission runs n=%d", n)
+        with transaction.atomic():
+            for run in SubmissionRun.objects.select_for_update().filter(
+                status=Status.PENDING_REMOTE
+            ):
+                if run.is_public:
+                    logger.debug("Added run=%d", run.id)
+                    set_submission_run_status(run, Status.PENDING)
+                    rt.submit_submission_run(client, run)
+
+                else:  # run is private
+                    for pair in SubmissionRunPair.objects.filter(private_run=run):
```

Comment:
This looks like it will work, but can you clarify why we can't just build a task graph off of the SubmissionRunPair object?